src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1914,6 +1914,32 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
             }
             tmp = ASR::make_ConstantLogical_t(al, x.base.base.loc, result, type);
             return;
+        } else if (call_name == "int") {
+            if (args.size() != 1) {
+                throw SemanticError(call_name + "() takes exactly one argument (" +
+                    std::to_string(args.size()) + " given)", x.base.base.loc);
+            }
+            ASR::expr_t* int_expr = args[0];
+            ASR::ttype_t* int_type = ASRUtils::expr_type(int_expr);
+            int int_kind = ASRUtils::extract_kind_from_ttype_t(int_type);
+            if (ASRUtils::is_integer(*int_type)) {
+                int64_t ival = ASR::down_cast<ASR::ConstantInteger_t>(ASRUtils::expr_value(int_expr))->m_n;
+                tmp = ASR::make_ConstantInteger_t(al, x.base.base.loc, ival, int_type);
+
+            } else if (ASRUtils::is_real(*int_type)) {
+                if (int_kind == 4) {
+                    float rv = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(int_expr))->m_r;
+                    int64_t ival = static_cast<int64_t>(rv);
+                    tmp = ASR::make_ConstantInteger_t(al, x.base.base.loc, ival, int_type);
+                } else {
+                    double rv = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(int_expr))->m_r;
+                    int64_t ival = static_cast<int64_t>(rv);
+                    tmp = ASR::make_ConstantInteger_t(al, x.base.base.loc, ival, int_type);
+                }
```

Comment:
This will work for compile time constants, but not for runtime variables. However, this can already be implemented: just cast real to an integer here using implicit cast.

Also, tests should be added.

Reply:
Done!

Reply:
> This will work for compile time constants, but not for runtime variables. However, this can already be implemented: just cast real to an integer here using implicit cast.

I tried this diff:
```diff
--- a/src/lpython/semantics/python_ast_to_asr.cpp
+++ b/src/lpython/semantics/python_ast_to_asr.cpp
@@ -1935,14 +1935,17 @@ public:

             } else if (ASRUtils::is_real(*int_type)) {
                 if (int_kind == 4) {
-                    float rv = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(int_expr))->m_r;
-                    int64_t ival = static_cast<int64_t>(rv);
-                    tmp = ASR::make_ConstantInteger_t(al, x.base.base.loc, ival, int_type);
+                    int_expr = ASR::down_cast<ASR::expr_t>(ASR::make_ImplicitCast_t(
+                            al, int_expr->base.loc, int_expr, ASR::cast_kindType::RealToInteger,
+                            ASRUtils::TYPE(ASR::make_Integer_t(al, x.base.base.loc, 4, nullptr, 0)), nullptr));
                 } else {
-                    double rv = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(int_expr))->m_r;
-                    int64_t ival = static_cast<int64_t>(rv);
-                    tmp = ASR::make_ConstantInteger_t(al, x.base.base.loc, ival, int_type);
+                    int_expr = ASR::down_cast<ASR::expr_t>(ASR::make_ImplicitCast_t(
+                            al, int_expr->base.loc, int_expr, ASR::cast_kindType::RealToInteger,
+                            ASRUtils::TYPE(ASR::make_Integer_t(al, x.base.base.loc, 8, nullptr, 0)), nullptr));
                 }
+                int64_t ival = ASR::down_cast<ASR::ConstantInteger_t>(ASRUtils::expr_value(int_expr))->m_n;
+                tmp = ASR::make_ConstantInteger_t(al, x.base.base.loc, ival, int_type);
+
             } else if (ASRUtils::is_logical(*int_type)) {
                 bool rv = ASR::down_cast<ASR::ConstantLogical_t>(int_expr)->m_value;
                 int8_t val = rv ? 1 : 0;
```
But I get this compiler error:
```bash
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
[2m  File "[0m[1m[35m$DIR/src/bin/lpython.cpp[39m[0m[2m", line 1470[0m
    return emit_asr(arg_file, runtime_library_dir,
[2m  File "[0m[1m[35m$DIR/src/bin/lpython.cpp[39m[0m[2m", line 547[0m
    r = LFortran::Python::python_ast_to_asr(al, *ast, diagnostics, true);
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 2029[0m
    auto res2 = body_visitor(al, *ast_m, diagnostics, unit, main_module);
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 1985[0m
    b.visit_Module(ast);
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 561[0m
    visit_stmt(*x.m_body[i]);
[2m  File "[0m[1m[35m$DIR/src/lpython/python_ast.h[39m[0m[2m", line 1827[0m
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
[2m  File "[0m[1m[35m$DIR/src/lpython/python_ast.h[39m[0m[2m", line 1698[0m
    case stmtType::FunctionDef: { v.visit_FunctionDef((const FunctionDef_t &)x); return; }
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 588[0m
    handle_fn(x, *ASR::down_cast<ASR::Subroutine_t>(t));
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 579[0m
    transform_stmts(body, x.n_body, x.m_body);
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 536[0m
    this->visit_stmt(*m_body[i]);
[2m  File "[0m[1m[35m$DIR/src/lpython/python_ast.h[39m[0m[2m", line 1827[0m
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
[2m  File "[0m[1m[35m$DIR/src/lpython/python_ast.h[39m[0m[2m", line 1702[0m
    case stmtType::Delete: { v.visit_Delete((const Delete_t &)x); return; }
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 682[0m
    this->visit_expr(*x.m_value);
[2m  File "[0m[1m[35m$DIR/src/lpython/python_ast.h[39m[0m[2m", line 1854[0m
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
[2m  File "[0m[1m[35m$DIR/src/lpython/python_ast.h[39m[0m[2m", line 1746[0m
    case exprType::Compare: { v.visit_Compare((const Compare_t &)x); return; }
[2m  File "[0m[1m[35m$DIR/src/lpython/semantics/python_ast_to_asr.cpp[39m[0m[2m", line 1946[0m
    int64_t ival = ASR::down_cast<ASR::ConstantInteger_t>(ASRUtils::expr_value(int_expr))->m_n;
[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 39[0m
    LFORTRAN_ASSERT(f != nullptr);
AssertFailed: f != nullptr
```

Reply:
The `down_cast` in `int64_t ival = ASR::down_cast<ASR::ConstantInteger_t>(ASRUtils::expr_value(int_expr))->m_n;` failed, meaning the `value` is not a `ConstantInteger`.